File: Data Structures & Algorithms/jump-game-ii/submission-1.go
func jump(nums []int) int {
    // similar to jump game 1
	// but here we have to return min number of steps 
	// and last index is always reachable

	// let's say from i index we can reach till j so we will always jump to
	// index between i..j from where we can reach till max index 
	// max(nums[i + 1] + i + 1, nums[i + 2] + i + 2, ... nums[i + j] + i + j) 
	

	ans := 0
	canReach := 0
	preReach := 0
	for i := range nums {
		if i == len(nums) - 1 {
			return ans
		}
		canReach = max(canReach, i + nums[i])	
		if i == preReach {
			preReach = canReach
			ans++
		}
	}
	return 0
}

File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int l = 0, r = heights.size() - 1;
        int ans = 0;
        while(l < r) {
            int water = min(heights[l], heights[r]) * (r - l);
            ans = max(ans, water);
            if (heights[r] < heights[l]) r--;
            else l++;
        }
        return ans;
    }
};
